### PR TITLE
Add various to helm create

### PR DESCRIPTION
**Pull request by naseemkullah**
**What this PR does / why we need it**:
Modifies the `helm create` command as follows:
 * ~Moves all deployment related values to an `.Values.deployment` umbrella as this makes more sense. (The same way ingress and service values are defined.)~  <- vetoed
* Adds an easy way to create an hpa for the deployment if the user so chooses to do so
* Adds an easy way to add environment variables for the deployment if the user so chooses to do so
* ~Makes deployment's probe paths/ports configurable~ no need really in retrospect, to have them as a value or directly done in the template, no difference for starter chart
* ~Adds an easy way to add security contexts for deployments, be it for pod or container~ Already tackled in seperate PR
* Adds an easy way to add volumes for deployments

**Special notes for your reviewer**:
Feedback is most welcome.

**If applicable**:
- [x] this PR contains documentation
- [x] this PR contains unit tests
- [x] this PR has been tested for backwards compatibility

### CODE REVIEW

**Comment by naseemkullah** [comment]
bump

**Comment by bacongobbler** [comment]
Because this changes the "best practices" for a starter chart (`helm create`), someone from the charts team may be more equipped to review this one. cc @helm/charts-maintainers

**Comment by unguiculus** [comment]
I'm not sure I like the additional level of nesting, especially since it's of technical nature. Usually, apps group by component if there are multiple.

**Comment by naseemkullah** [comment]
> I'm not sure I like the additional level of nesting, especially since it's of technical nature. Usually, apps group by component if there are multiple.

I see, I see.. do you mean like this? 
https://github.com/helm/charts/blob/master/stable/nginx-ingress/values.yaml

**Comment by naseemkullah** [comment]
@unguiculus I've reverted the additional nesting, what about the other stuff, anything you like/don't like?

**Comment by naseemkullah** [comment]
Should I change `.Values.volumes`,`.Values.volumeMounts` and `.Values.env` for `.Values.extraVolumes`,`.Values.extraVolumeMounts` and `.Values.extraEnvVars` respectively, or something of the sort?

**Comment by naseemkullah** [comment]
bump

**Comment by naseemkullah** [comment]
bump

**Comment by naseemkullah** [comment]
bump

**Comment by naseemkullah** [comment]
Hello?

**Comment by naseemkullah** [comment]
bump (please let me know if there is anything worth keeping, if anything just the hpa stuff maybe??)

File: pkg/chartutil/create.go [hunk 1]
```diff
@@ -96,6 +104,40 @@ resources: {}
...
+  # runAsNonRoot: true
+  # runAsUser: 10001
+
+env: []
+  # - name: FOO
+  #   value: bar
+  # - name: BAZ
+  #   value: qux
+
+readinessProbe:
```
Comment (bacongobbler) [inline pkg/chartutil/create.go]: I don't think a readinessProbe should be enabled by default. Not all applications respond to HTTP requests (in fact, Tiller's one such example), and not all of them respond with a 200 OK on the root URL.

File: pkg/chartutil/create.go [hunk 2]
```diff
@@ -96,6 +104,40 @@ resources: {}
...
+  httpGet:
+    path: /
+    port: http
+livenessProbe:
```
Comment (bacongobbler) [inline pkg/chartutil/create.go]: Same comment here. I don't think a livenessProbe should be enabled by default. Not all applications respond to HTTP requests (in fact, Tiller's one such example), and not all of them respond with a 200 OK on the root URL.

**Review by bacongobbler — CHANGES_REQUESTED** [review]
Thank you for the contribution.

As a refresher, `helm create` was originally intended as a command to scaffold a chart directory with the right layout, along with a very basic "getting started" example. It's very similar to `hugo new`, `rails new`, or `django-admin startproject`; It's intended to be a great "out of the box" experience providing some sane defaults, but getting out of the way for the more advanced users who want to extend upon the basics.

With that in mind, `helm create` was designed to create a chart that can be immediately installable with sane defaults, providing users a good "head start" to write their first chart.

Makes sense?

As a user, I should expect that any and all templates and values should be documented, including

- who is this for?
- what does it do?
- where is this useful?
- when enabled, how is my application affected?
- why should I use this?
- how I can consume/configure these values for my application?

I'm curious to know why there's no supporting documentation provided with the values added with this PR. All of the values provided have been disabled by default (except the liveness/readiness probes, which I left additional feedback below), and as far as I can tell, all of the questions provided above were left unanswered, leaving me confused.

That's the main issue I have with this PR: the "best practices" defined in this PR are all undocumented with no information on how it can be used by first-timers and advanced maintainers alike. The horizontal pod autoscaler, security contexts, volumes, etc. are disabled by default, and it's up to the user to figure out how to use them with no guidance provided. This ends up confusing (and ultimately frustrating) both first-timers and more knowledgeable chart maintainers.

The other issue I have with this PR: all of the values here are disabled by default, and are only useful in certain use cases, but don't always apply. Many applications don't rely on volumes for example. Making these part of the default starter chart only adds more cognitive overhead to the user, as they now have to determine if they actually need to add a horizontal pod autoscaler or volume to their application, and if they don't, how they can remove these "useless" templates from the starter chart.

If you still want to move forward with this, my suggestion would be to start from scratch. Remove all of the values/templates provided here, and start by writing documentation on how users can add horizontal pod autoscalers/security contexts/volumes to their charts under the "Developing Charts" guide in the docs and answer the "5 W's" provided above.

What do you think about that approach?

If that does not work for you, this may be a good [starter chart](https://docs.helm.sh/developing_charts/#chart-starter-packs) similar to [the incubator/common chart](https://github.com/helm/charts/tree/master/incubator/common). Given that a large amount of the features provided here are only applicable in certain scenarios, closing this PR and going forward with making this part of an opinionated starter chart may be a better long-term solution.

Let me know what you think!

File: pkg/chartutil/create.go (hunk 1 above)
Comment (naseemkullah) [reply pkg/chartutil/create.go]: true I was just following what comes out of the box with helm create atm

File: pkg/chartutil/create.go (hunk 2 above)
Comment (naseemkullah) [reply pkg/chartutil/create.go]: true, just following what comes with helm create atm.
I can tweak this (and readiness) to be disabled by default. 

Should these also be made configurable to be something else than an `httpGet` ?

Comment (bacongobbler) [reply pkg/chartutil/create.go]: I was suggesting that this should all be removed completely. Did you read my earlier comment?

Comment (naseemkullah) [reply pkg/chartutil/create.go]: I did, I did. I'm sorry I don't see where you said to remove the probes completely.

**Comment by naseemkullah** [comment]
Arguably should we have a deployment template? an ingress template? some charts are just a cronjob... I was just building on what the helm create does now, it creates a typical web api workload. HPA is very useful for this kind of workload

**Comment by naseemkullah** [comment]
Closing. Thanks for feedback everybody.

**Pull request closed without merging**